API doc directives: accept full character range in package names

poWer4aiX opened the pull request:
According to the [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se8/html/index.html) member of packages, e.g. subpackages are identifiers. Identifiers itself is

>  ...an unlimited-length sequence of Java letters and Java digits, the first of which must be a Java letter. 

As a `Java letter` includes more than what is expressed by the regex `\b[a-z][a-z0-9_]*)\.`, package names which include uppercase or non ASCII characters cannot be processed. I know that there are some _naming conventions_ flying around but something like `aBc.DE.fg` is still a valid package name in my opinion. 

I have changed the regex in `packageDotsToSlash()` to `(\b\p{javaJavaIdentifierStart}\p{javaJavaIdentifierPart}*)\.`. Unfortunately this breaks with the changes applied for issue #397, #395, #98, #86 to handle inner classes. In order to support this notation, the regex can be relaxed to `(\b\p{javaLowerCase}\p{javaJavaIdentifierPart}*)\.`.

I have added the variable 'scaladoc.strictPackageIdent' (which defaults to `false`) by which you can switch between the two regexp. Once set to true, the package name can conform to an identifier now.

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

poWer4aiX commented:
Stupid question, is there anything I have to do in order to pass the check? I don't understand why "the committer (me) should not linked to a user"...

lightbend-cla-validator commented:
Hi @poWer4aiX,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

poWer4aiX commented:
CLA signed and replied by email...

raboof commented:
Your CLA signature somehow didn't make it into our system yet. That's probably a problem on our side - how did you submit it?

poWer4aiX commented:
Hi,

second try by using "reply all". Simply replying, resulted i a mail send
to "notifications@github.com", which seems to be wrong.

Thanks,
Christian

Am 21.04.2020 um 20:21 schrieb Lightbend CLA Validator:
>
> Hi @poWer4aiX <https://github.com/poWer4aiX>,
>
> Thank you for your contribution! We really value the time you've taken
> to put this together.
>
> Before we proceed with reviewing this pull request, please sign the
> Lightbend Contributors License Agreement:
>
> http://www.lightbend.com/contribute/cla
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightbend/paradox/pull/431#issuecomment-617332363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTO6JCPOAA366PJMHYUJNDRNXPZVANCNFSM4MNK5PWA>.
>

poWer4aiX commented:
@raboof, I simply filled out the PDF and attached it by replying to the email I received. Just noticed that this was addressed to `notifications@github.com` which is obviously the wrong address, I guess. 

Resent it some minutes ago to `reply+...`, hope this is working now. Sorry for any inconveniences.

lightbend-cla-validator commented:
Hi @poWer4aiX,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

ennru commented:
Your CLA still hasn't shown up in our system, I'm sorry.
Can you try and sign it with the Github integration?

poWer4aiX commented:
Can you please explain how to "sign it with the Github integration" or provide me a link with a description? BTW, I just sent it to "cla@lightbend.com".

ennru commented:
Sorry for all this trouble, on https://www.lightbend.com/contribute/cla you'll find the "Confirm with your GitHub account" link.

ennru commented:
Ok, all good now. I checked the CLA manually.

poWer4aiX commented:
Thanks for your support.

Well in the latest version I have applied the change for `@scaladoc` and `@javadoc`. In addition I followed I added a setting for `[javadoc|scaladoc].<package-prefix>.package_name_style`, similar to `jadadoc...link_style`. Defaults to `startWithLowercase`, can be set to `startWithAnycase`.

raboof commented on core/src/main/scala/com/lightbend/paradox/markdown/Directive.scala:
```diff
@@ -311,25 +311,53 @@ abstract class ApiDocDirective(name: String)
+   * Converts package dot notation to a path, separated by '/'
+   * Allow all valid java characters and java numbers to be used, according to the java lang spec. In order to
+   * be allow to convert `package.Outer.Inner` in scala to the `packge/Outer$$Inner`, `strictPackageIdent` has
+   * be set to `true`.
```
Seems like this is no longer accurate. Good to give an example: `startWithLowercase` will get it wrong when a package name starts with an uppercase letter or when an inner class starts with a lowercase character, while `startWithAnycase` will derive the wrong path whenever an inner class is encountered. Does that summarize the trade-off correctly?

poWer4aiX replied:
You're right, it's the old style of setting. Will shorten this block and add your description to the `@param` description.

```suggestion
   * Converts package dot notation to a path, separated by '/'
   * Allow all valid java characters and java numbers to be used, according to the java lang spec.
   *
   * @param s package or full qualified class name to be converted.
   * @param packageNameStyle Setting `startWithLowercase`` will get it wrong when a package name
   *                         starts with an uppercase letter or when an inner class starts with
   *                         a lowercase character, while `startWithAnycase` will derive the wrong
   *                         path whenever an inner class is encountered.
```

raboof commented on core/src/main/scala/com/lightbend/paradox/markdown/Directive.scala:
```diff
@@ -311,25 +311,53 @@ abstract class ApiDocDirective(name: String)
+   *                          `startWithAnycase` allows packages to start with uppercase also.
+   * @return Resulting path.
+   */
+  def packageDotsToSlash(s: String, packageNameStyle: String = "startWithLowercase") =
```
This default isn't used anywhere, right?
```suggestion
  def packageDotsToSlash(s: String, packageNameStyle: String) =
```

poWer4aiX replied:
Yes it is not explicitly checked. In the first thought I wanted to define it as simple flag, but using it as a style is more flexile. Was also wondering if it makes sense to add the old style where ascii characters were allowed only, but skipped this. Do you think it is worse to add?

raboof replied:
I don't think we need the old behavior anymore :+1:

raboof commented on core/src/main/scala/com/lightbend/paradox/markdown/Directive.scala:
```diff
@@ -311,25 +311,53 @@ abstract class ApiDocDirective(name: String)
 
+  import ScaladocDirective._
+
+  val defaultPackageNameStyle = variables.getOrElse("scaladoc.package_name_style", "startWithLowercase")
```
:+1: to keep the default

poWer4aiX commented:
If you think the change in general is fine, I will also update the documentation.

poWer4aiX commented:
@raboof in `@javadoc` the cases with inner classes

 - having a subpackage starting with an uppercase character (org.example.Lib.Outer.Inner)
 - having an outer class starting with a lowercase character (org.example.lib.outer.Inner)
 - having the inner class starting with a lowercase character (org.example.lib.Outer.inner)

are not addressed so far. `@javadoc` will resolve to an incorrect URL in all cases. To fix this I propose to make use of the opposite from scala and resolve `$$` in the FQCN to a `.`.

I have implemented it and added the corresponding tests as well.

raboof commented on core/src/main/scala/com/lightbend/paradox/markdown/Directive.scala:
```diff
@@ -371,14 +391,31 @@ case class JavadocDirective(ctx: Writer.Context)
+
+  private[markdown] def url(link: String, baseUrl: Url, linkStyle: LinkStyle, packageNameStyle: String): Url = {
+    val url = Url(link).base
+    val path = dollarDollarToClassDot(ApiDocDirective.packageDotsToSlash(url.getPath, packageNameStyle)) + ".html"
```
So any names *actually* containing `$$` are now broken - I suppose that should be exotic enough that this is OK.

poWer4aiX replied:
Yes I also think that this might be a very theoretical case. If we want to get rid of it, we can also think of using a non java character, e.g. a `..` (two dots) instead of `$$`. If I', right then this is invalid for any full qualified name.

mdedetrich commented:
Is there a plan to merge this PR? This is mainly in relation to https://github.com/lightbend/paradox/issues/515 but in general its been approved but sitting here unmerged for 2 years.

mdedetrich commented:
Perfect, thank you!
